File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/AsyncOpMode.java
package org.firstinspires.ftc.teamcode.utils;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.pedropathing.follower.Follower;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import org.firstinspires.ftc.teamcode.RobotHardware;
import org.firstinspires.ftc.teamcode.intake.IntakePositions;
import org.firstinspires.ftc.teamcode.outtake.OuttakePositions;
import org.firstinspires.ftc.teamcode.utils.systems.CancelableAction;

import java.util.ArrayList;
import java.util.List;

public abstract class AsyncOpMode extends LinearOpMode {
    /**
     * The robot hardware.
     */
    protected RobotHardware robot;
    /**
     * The Pedro follower.
     */
    protected Follower follower;
    /**
     * The list of systems that need updating.
     */
    protected List<CancelableAction> actions;
    /**
     * The delayed actions.
     */
    private final DelayedActions delayed = new DelayedActions();
    /**
     * Adds an action to be run after a delay.
     * @param delay the delay in seconds.
     * @param action the action to run.
     */
    protected void addDelay(double delay, Runnable action) {
        delayed.addDelayed(delay, action);
    }
    /**
     * Init of the opMode
     */
    public abstract void systemInit();
    /**
     * Loop for the init of the opMode (if needed)
     */
    public void systemInitLoop() {

    }
    /**
     * Start of the opMode (after waitForStart(), if needed)
     */
    public void systemStart() {

    }
    /**
     * Loop for the opMode
     */
    public abstract void systemLoop();
    @Override
    public void runOpMode() {
        robot = new RobotHardware(hardwareMap);
        FtcDashboard dashboard = FtcDashboard.getInstance();
        follower = new Follower(hardwareMap);
        actions = new ArrayList<>();
        systemInit();
        while (!isStarted()) {
            TelemetryPacket packet = new TelemetryPacket();
            actions.removeIf(action -> !action.run(packet));
            dashboard.sendTelemetryPacket(packet);
            systemInitLoop();
        }
        dashboard.clearTelemetry();
        systemStart();
        while (opModeIsActive()) {
            delayed.run();
            TelemetryPacket packet = new TelemetryPacket();
            packet.put("Robot Pose", follower.getPose().getAsPedroCoordinates());
            actions.removeIf(action -> !action.run(packet));
            dashboard.sendTelemetryPacket(packet);
            follower.update();
            follower.drawOnDashBoard();
            systemLoop();
        }
    }

    protected void transfer(boolean auto) {
        robot.outtake.setClaw(false);
        if (robot.intake.getTargetPosition() == IntakePositions.PICKUP) {
            robot.intake.setTargetPosition(IntakePositions.TRANSFER);
            addDelay(0.2, () -> robot.outtake.setTargetPosition(OuttakePositions.TRANSFER));
        } else robot.outtake.setTargetPosition(OuttakePositions.TRANSFER);
        if (auto) addDelay(0.5, () -> {
            if (robot.outtake.isClosed()) {
                robot.intake.setClaw(true);
                addDelay(0.1, () -> robot.outtake.setClaw(false));
            } else {
                robot.outtake.setClaw(true);
                addDelay(0.1, () -> robot.intake.setClaw(false));
            }
        });
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/utils/AsyncOpMode.kt
package org.firstinspires.ftc.utils

import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.pedropathing.follower.Follower
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.RobotHardware
import org.firstinspires.ftc.utils.systems.CancelableAction

abstract class AsyncOpMode : LinearOpMode() {
    /**
     * The hardware for the robot.
     */
    protected lateinit var robot: RobotHardware
    /**
     * Pedro Pathing`s follower
     */
    protected lateinit var follower: Follower
    /**
     * Creates a list with the system or subsystem
     */
    private val actions: MutableList<CancelableAction> = mutableListOf()
    /**
     * Creates the container for the delayed
     */
    private val delayed = DelayedActions()
    /**
     * Creates the container for the checks
     */
    private val checks = CheckActions()
    /**
     * A wrapper function for addDelay in [DelayedActions]
     * @see DelayedActions.addDelay
     * @param time the delay in seconds.
     * @param action the action to run.
     */
    protected fun delay(time: Long, action: () -> Unit) {
        delayed.addDelay(time, action)
    }
    /**
     * A wrapper function for addCheck in [CheckActions]
     * @see CheckActions.addCheck
     * @param check the lambda check that determines if the action should be run.
     * @param action the action to run.
     */
    protected fun check(check: () -> Boolean, action: () -> Unit) {
        checks.addCheck(check, action)
    }
    /**
     * Init of the opMode
     */
    abstract fun systemInit()
    /**
     * Loop to be run before the start of the opMode
     */
    open fun systemInitLoop() {}
    /**
     * Start of the opMode
     */
    open fun systemStart() {}
    /**
     * Loop to be run after the start of the opMode
     */
    abstract fun systemLoop()

    final override fun runOpMode() {
        // Initializes the robot hardware
        robot = RobotHardware(hardwareMap)
        // Initializes the dashboard
        val dashboard = FtcDashboard.getInstance()
        // Initializes the follower
        follower = Follower(hardwareMap)
        systemInit()
        while (!isStarted) {
            // Runs the systems while in init
            val packet = TelemetryPacket()
            actions.removeAll { !it.run(packet) }
            dashboard.telemetry.update()
            systemInitLoop()
        }
        dashboard.clearTelemetry()
        systemStart()
        while (opModeIsActive()) {
            // Runs the systems while in play
            val packet = TelemetryPacket()
            actions.removeAll { !it.run(packet) }
            delayed.run()
            checks.run()
            // Updates the follower and dashboard
            packet.put("Robot Pose", follower.pose.asPedroCoordinates)
            dashboard.telemetry.update()
            follower.drawOnDashBoard()
            follower.update()
            systemLoop()
        }
    }

    /**
     * Adds timed transfer logic to the opMode.
     * Only use in auto.
     */
    protected fun transfer() {
        //TODO: Do after rest of the systems are done
    }
}